libkbfs/block_retrieval_queue.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"container/heap"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultBlockRetrievalWorkerQueueSize int = 100
+	defaultOnDemandRequestPriority       int = 100
+)
+
+// blockRetrievalRequest represents one consumer's request for a block.
+type blockRetrievalRequest struct {
+	ctx    context.Context
+	block  Block
+	doneCh chan error
+}
+
+// blockRetrieval contains the metadata for a given block retrieval. May
+// represent many requests, all of which will be handled at once.
+type blockRetrieval struct {
+	// The block being retrieved
+	blockPtr BlockPointer
+	// The key metadata for the request
+	kmd KeyMetadata
+	// The index of the retrieval in the heap
+	index int
+	// The priority of the retrieval
+	priority int
+	// The count of elements in the heap when this element was inserted.
```

Comment:
This doesn't seem quite right -- it sounds here like you're talking about the current length of the heap, but I think you really mean the total number of requests that have ever been queued,